File: wafel_py_old/ui/joystick_control.py
from typing import *
from dataclasses import dataclass
import math

import wafel.imgui as ig
from wafel.local_state import use_state


@dataclass
class JoystickControlState:
  start_value: Optional[Tuple[float, float]] = None

  def get_value(self, drag: Tuple[float, float]) -> Tuple[float, float]:
    assert self.start_value is not None
    return (self.start_value[0] + drag[0], self.start_value[1] + drag[1])

  @property
  def active(self) -> bool:
    return self.start_value is not None

  def set_active(self, value: Tuple[float, float]) -> None:
    if self.start_value is None:
      self.start_value = value

  def reset(self) -> None:
    self.start_value = None


def render_joystick_control(
  id: str,
  stick_x: float,
  stick_y: float,
  shape = 'square',
) -> Optional[Tuple[float, float]]:
  ig.push_id(id)
  state = use_state('', JoystickControlState()).value

  dl = ig.get_window_draw_list()

  padding = 10
  content_region = ig.get_content_region_available()
  size = min(
    content_region.x - ig.get_style().scrollbar_size - 2 * padding,
    content_region.y - 2 * padding,
    200,
  )
  size = max(size, 100)

  initial_cursor_pos = ig.get_cursor_pos()
  top_left = (
    initial_cursor_pos[0] + ig.get_window_position()[0] - ig.get_scroll_x() + padding,
    initial_cursor_pos[1] + ig.get_window_position()[1] - ig.get_scroll_y() + padding,
  )

  background_color = ig.get_color_u32_rgba(0, 0, 0, 0.3)
  if shape == 'square':
    dl.add_rect_filled(
      top_left[0],
      top_left[1],
      top_left[0] + size,
      top_left[1] + size,
      background_color,
    )
  elif shape == 'circle':
    dl.add_circle_filled(
      top_left[0] + size / 2,
      top_left[1] + size / 2,
      size / 2,
      background_color,
      num_segments = 32,
    )

  result = None

  if state.active and ig.is_mouse_down():
    new_offset = state.get_value(ig.get_mouse_drag_delta(lock_threshold=0))

    new_stick_x = new_offset[0] / size * 2 - 1
    new_stick_y = (1 - new_offset[1] / size) * 2 - 1
    if shape == 'square':
      new_stick_x = min(max(new_stick_x, -1), 1)
      new_stick_y = min(max(new_stick_y, -1), 1)
    elif shape == 'circle':
      mag = math.sqrt(new_stick_x**2 + new_stick_y**2)
      if mag > 1:
        new_stick_x /= mag
        new_stick_y /= mag

    if (new_stick_x, new_stick_y) != (stick_x, stick_y):
      stick_x, stick_y = new_stick_x, new_stick_y
      result = (stick_x, stick_y)

  offset = (
    (stick_x + 1) / 2 * size,
    (1 - (stick_y + 1) / 2) * size,
  )

  dl.add_line(
    top_left[0] + size / 2,
    top_left[1] + size / 2,
    top_left[0] + offset[0],
    top_left[1] + offset[1],
    ig.get_color_u32_rgba(1, 1, 1, 0.5),
  )

  button_size = 20
  button_pos = (
    padding + initial_cursor_pos[0] + offset[0] - button_size / 2,
    padding + initial_cursor_pos[1] + offset[1] - button_size / 2,
  )
  ig.set_cursor_pos(button_pos)
  ig.button('##joystick-button', button_size, button_size)

  ig.set_cursor_pos((
    initial_cursor_pos[0],
    initial_cursor_pos[1] + size + 2 * padding,
  ))

  if ig.is_item_active():
    state.set_active(offset)
  else:
    state.reset()

  ig.pop_id()
  return result


__all__ = ['render_joystick_control']

File: wafel_app/src/joystick_control.rs
use std::sync::Mutex;

use imgui::{self as ig, im_str};
use once_cell::sync::OnceCell;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub(crate) enum JoystickControlShape {
    Square,
    Circle,
}

#[derive(Debug)]
struct ActiveJoystickState {
    id: ig::sys::ImGuiID,
    start_value: [f32; 2],
}

fn active_joystick_state() -> &'static Mutex<Option<ActiveJoystickState>> {
    static INSTANCE: OnceCell<Mutex<Option<ActiveJoystickState>>> = OnceCell::new();
    INSTANCE.get_or_init(|| Mutex::new(None))
}

pub(crate) fn render_joystick_control(
    ui: &ig::Ui<'_>,
    id: &str,
    mut stick: [f32; 2],
    shape: JoystickControlShape,
) -> Option<[f32; 2]> {
    let id_token = ui.push_id(id);

    let state_id = unsafe { ig::sys::igGetID_Str(im_str!("state").as_ptr()) };
    let mut global_state = active_joystick_state().lock().unwrap();

    let dl = ui.get_window_draw_list();

    let padding = 10.0;
    let content_region = ui.content_region_avail();
    let size = (content_region[0] - ui.clone_style().scrollbar_size - 2.0 * padding)
        .min(content_region[1] - 2.0 * padding)
        .min(200.0)
        .max(100.0);

    let initial_cursor_pos = ui.cursor_pos();
    let top_left = [
        initial_cursor_pos[0] + ui.window_pos()[0] - ui.scroll_x() + padding,
        initial_cursor_pos[1] + ui.window_pos()[1] - ui.scroll_y() + padding,
    ];

    let background_color = ig::ImColor32::from_rgba_f32s(0.0, 0.0, 0.0, 0.3);
    match shape {
        JoystickControlShape::Square => dl
            .add_rect(
                top_left,
                [top_left[0] + size, top_left[1] + size],
                background_color,
            )
            .filled(true)
            .build(),
        JoystickControlShape::Circle => dl
            .add_circle(
                [top_left[0] + size * 0.5, top_left[1] + size * 0.5],
                size * 0.5,
                background_color,
            )
            .num_segments(32)
            .filled(true)
            .build(),
    }

    let mut result = None;

    if let Some(state) = global_state.as_ref().filter(|state| state.id == state_id) {
        if ui.is_mouse_down(ig::MouseButton::Left) {
            let delta = ui.mouse_drag_delta_with_threshold(ig::MouseButton::Left, 0.0);
            let new_offset = [
                state.start_value[0] + delta[0],
                state.start_value[1] + delta[1],
            ];

            let mut new_stick = [
                new_offset[0] / size * 2.0 - 1.0,
                (1.0 - new_offset[1] / size) * 2.0 - 1.0,
            ];

            match shape {
                JoystickControlShape::Square => {
                    new_stick[0] = new_stick[0].clamp(-1.0, 1.0);
                    new_stick[1] = new_stick[1].clamp(-1.0, 1.0);
                }
                JoystickControlShape::Circle => {
                    let mag = (new_stick[0].powi(2) + new_stick[1].powi(2)).sqrt();
                    if mag > 1.0 {
                        new_stick[0] /= mag;
                        new_stick[1] /= mag;
                    }
                }
            }

            #[allow(clippy::float_cmp)]
            if new_stick != stick {
                stick = new_stick;
                result = Some(stick);
            }
        }
    }

    let offset = [
        (stick[0] + 1.0) / 2.0 * size,
        (1.0 - (stick[1] + 1.0) / 2.0) * size,
    ];

    dl.add_line(
        [top_left[0] + size / 2.0, top_left[1] + size / 2.0],
        [top_left[0] + offset[0], top_left[1] + offset[1]],
        ig::ImColor32::from_rgba_f32s(1.0, 1.0, 1.0, 0.5),
    )
    .build();

    let button_size = 20.0;
    let button_pos = [
        padding + initial_cursor_pos[0] + offset[0] - button_size / 2.0,
        padding + initial_cursor_pos[1] + offset[1] - button_size / 2.0,
    ];
    ui.set_cursor_pos(button_pos);
    ui.button_with_size(im_str!("##joystick-button"), [button_size, button_size]);

    ui.set_cursor_pos([
        initial_cursor_pos[0],
        initial_cursor_pos[1] + size + 2.0 * padding,
    ]);

    let already_active = global_state
        .as_ref()
        .filter(|state| state.id == state_id)
        .is_some();
    if ui.is_item_active() {
        if !already_active {
            *global_state = Some(ActiveJoystickState {
                id: state_id,
                start_value: offset,
            });
        }
    } else if already_active {
        *global_state = None;
    }

    id_token.pop();
    result
}
